odoo/orm/query.py

```diff
@@ -270,3 +272,95 @@ def __len__(self) -> int:
 
     def __iter__(self) -> Iterator[int]:
         return iter(self.get_result_ids())
+
+
+class TableSQL(SQL):
+    """An alias of a table in a Query."""
+    __slots__ = ('_alias', '_model', '_query')
+
+    def __init__(self, alias: str, model: BaseModel | None, query: Query):
+        assert isinstance(alias, str)
+        self._alias = alias
+        self._model = model
+        self._query = query
+
+    @property
+    def _sql_tuple(self):
+        return SQL.identifier(self._alias)._sql_tuple
+
+    def __getitem__(self, field_name: str) -> SQL:
+        if (model := self._model) is None:
+            return SQL.identifier(self._alias, field_name)
+        field = model._fields.get(field_name)
+        if field is None:
+            raise ValueError(f"Invalid field {field_name!r} on model {model._name!r}")
+        return FieldSQL(self, field)
+
+    __getattr__ = __getitem__
+
+    def __repr__(self):
+        return f"TableSQL({self._alias!r}, {self._model._name if self._model else '-'}, {self._query!r})"
+
+    def _with_model(self, model: BaseModel) -> TableSQL:
+        """ Bind a model to a new instance. """
+        assert self._model is None or self._model._name == model._name
+        return TableSQL(self._alias, model, self._query)
+
+    def _sudo(self, flag: bool = True) -> TableSQL:
+        """ Like `_with_model`, just sudo existing model. """
+        if self._model is None or self._model.env.su == flag:
+            return self
+        return self._with_model(self._model.sudo(flag))
+
+    def _make_alias(self, link: str, model: BaseModel | None = None) -> TableSQL:
+        """ Generate a new table/alias using this as source. """
+        return TableSQL(_generate_table_alias(self._alias, link), model, self._query)
+
+    def _join(
+        self,
+        field_name: str,
+        **kw,
+    ) -> TableSQL:
+        """ Join a table.
+
+        Given a field, use its `Field.join` to join to another model.
+        """
+        model = self._model
+        if model is None:
+            raise ValueError(f"Cannot {self}._join() without a model")
+        field = model._fields.get(field_name)
+        if not field:
+            raise ValueError(f"Invalid field {field_name!r} on model {model._name!r}")
+        if field.type == 'many2one':
+            return field.join(self, **kw)
+        raise ValueError(f"Invalid field {field_name}: _join() not possible")
+
+
+class FieldSQL(SQL):
+    """ A Field and a TableSQL.
+
+    Accessing an attribute builds the SQL for the property of the field.
+    In case of many2one fields, we join the table and resolve the field on the
+    comodel.
```

Comment:
I would make a separate commit for this part, introducing the feature and defining `property_to_sql` on many2one fields (and improving the other implementations with better error messages).